TESTS/mbed_hal/watchdog/main.cpp

```diff
@@ -0,0 +1,125 @@
+/* mbed Microcontroller Library
+ * Copyright (c) 2017 ARM Limited
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#if !DEVICE_WATCHDOG
+#error [NOT_SUPPORTED] Watchdog not supported for this target
+#endif
+
+#include "greentea-client/test_env.h"
+#include "utest/utest.h"
+#include "unity/unity.h"
+#include "hal/watchdog_api.h"
+#include "watchdog_api_tests.h"
+
+using namespace utest::v1;
+
+#define TIMEOUT_DELTA_MS 5UL
+#define WDG_TIMEOUT_MS 500UL
+
+const watchdog_config_t WDG_CONFIG_DEFAULT = {
+    .timeout_ms = WDG_TIMEOUT_MS
+};
+
+void test_max_timeout_is_valid()
+{
+    TEST_ASSERT(hal_watchdog_get_platform_features().max_timeout > 1UL);
+}
+
+void test_restart_is_possible()
+{
+    watchdog_features_t features = hal_watchdog_get_platform_features();
+    if (!features.disable_watchdog) {
+        TEST_IGNORE_MESSAGE("Disabling watchdog not supported for this platform");
+        return;
+    }
+    TEST_ASSERT(features.update_config);
+}
+
+void test_stop()
+{
+    watchdog_features_t features = hal_watchdog_get_platform_features();
+    if (!features.disable_watchdog) {
+        TEST_ASSERT_EQUAL(WATCHDOG_STATUS_NOT_SUPPORTED, hal_watchdog_stop());
+        TEST_IGNORE_MESSAGE("Disabling watchdog not supported for this platform");
+        return;
+    }
+
+    TEST_ASSERT_EQUAL(WATCHDOG_STATUS_OK, hal_watchdog_stop());
+
+    TEST_ASSERT_EQUAL(WATCHDOG_STATUS_OK, hal_watchdog_init(&WDG_CONFIG_DEFAULT));
+    TEST_ASSERT_EQUAL(WATCHDOG_STATUS_OK, hal_watchdog_stop());
+    // Make sure that a disabled watchdog does not reset the core.
+    wait_ms(WDG_TIMEOUT_MS + TIMEOUT_DELTA_MS);
+
+    TEST_ASSERT_EQUAL(WATCHDOG_STATUS_OK, hal_watchdog_stop());
+}
+
+void test_update_config()
+{
+    watchdog_features_t features = hal_watchdog_get_platform_features();
+    if (!features.update_config) {
+        TEST_IGNORE_MESSAGE("Updating watchdog config not supported for this platform");
+        return;
+    }
+
+    watchdog_config_t config = WDG_CONFIG_DEFAULT;
+    TEST_ASSERT_EQUAL(WATCHDOG_STATUS_OK, hal_watchdog_init(&config));
+    TEST_ASSERT_EQUAL_UINT32(WDG_TIMEOUT_MS, hal_watchdog_get_reload_value());
+
+    config.timeout_ms = features.max_timeout - 1;
+    TEST_ASSERT_EQUAL(WATCHDOG_STATUS_OK, hal_watchdog_init(&config));
+    TEST_ASSERT_EQUAL_UINT32(features.max_timeout - 1, hal_watchdog_get_reload_value());
```

Comment:
After K64F watchdog stop problem got fixed 2e7ce31a9a834209e6ac7a4fa9a0ec8d9672f246 , this test still failed here on K64F, and got an error of:
```[1515493472.85][CONN][RXD] >>> Running case #4: 'Update config with multiple init calls'...
[1515493472.91][CONN][INF] found KV pair in stream: {{__testcase_start;Update config with multiple init calls}}, queued...
[1515493472.96][CONN][RXD] :83::FAIL: Expected 4294967287 Was 500
[1515493473.03][CONN][INF] found KV pair in stream: {{__testcase_finish;Update config with multiple init calls;0;1}}, queued...
[1515493473.12][CONN][RXD] >>> 'Update config with multiple init calls': 0 passed, 1 failed with reason 'Assertion Failed'```

Reply:
I'll look into it. But, I think it would be best if this function is removed. I can't see it being very useful as the user should know what timeout they set, and it may not return the same value as it's recalculating the timeout from the register*prescaler which may be slightly different to what they initially set.

Reply:
Doesn't K64F have some restriction on R/W access to watchdog control registers? Adding `wait_us(1);` in between `hal_watchdog_init(&config)` calls seem to help.

Reply:
That's most likely the cause, adding the wait is a reasonable solution.

Reply:
I think this should rather be buried inside a platform specific implementation. If K64F needs to do some waiting, then `hal_watchdog_init()` should not return until everything is ready. Alternatively, `hal_watchdog_init()` could perform some checks before doing the actual init. I asked @0xc0170 for a comment and he has a similar opinion.